src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilter.java

```diff
@@ -0,0 +1,330 @@
+            } else if (read.isSecondOfPair()) {
+                newRead.setIsSecondOfPair();
+            }
+        }
+        final String readGroup = read.getReadGroup();
+        if (readGroup != null) {
+            newRead.setAttribute(SAMTag.RG.name(), readGroup);
+        }
+        return newRead;
+    }
+
+    /**
+     * Sets proper pairedness flags
+     */
+    private static Iterator<GATKRead> setPartitionUnpairedFlags(final Iterator<GATKRead> iter, final int readsPerPartitionGuess) {
+        final Tuple2<List<GATKRead>, List<GATKRead>> lists = getPairedAndUnpairedLists(iter, readsPerPartitionGuess);
+        final List<GATKRead> pairedReadsList = lists._1;
+        final List<GATKRead> unpairedReadsList = lists._2;
+        for (final GATKRead unpairedRead : unpairedReadsList) {
+            unpairedRead.setIsPaired(false);
+        }
+        final List<GATKRead> newPartitionList = new ArrayList<>(pairedReadsList.size() + unpairedReadsList.size());
+        newPartitionList.addAll(pairedReadsList);
+        newPartitionList.addAll(unpairedReadsList);
+        return newPartitionList.iterator();
+    }
+
+    /**
+     * Repartitions reads so that reads with the same name will be on the same partition
+     */
+    static JavaRDD<GATKRead> repartitionPairedReads(final JavaRDD<GATKRead> reads) {
+        //Shuffle reads into partitions by read name hash code
+        return reads.mapToPair(read -> new Tuple2<>(read.getName(), read))
+                .partitionBy(new HashPartitioner(reads.getNumPartitions()))
+                .map(Tuple2::_2);
+    }
+
+    /**
+     * Maps partition to a Tuple of two Lists, the first containing the paired reads, the second containing unpaired
+     */
+    static Tuple2<List<GATKRead>, List<GATKRead>> getPairedAndUnpairedLists(final Iterator<GATKRead> iter, final int readsPerPartitionGuess) {
+        //Find the paired and unpaired reads by scanning the partition for repeated names
+        final List<GATKRead> pairedReadsList = new ArrayList<>(readsPerPartitionGuess);
+        final Map<String, GATKRead> unpairedReads = new HashMap<>(readsPerPartitionGuess);
+        while (iter.hasNext()) {
+            final GATKRead read = iter.next();
+            final String readName = read.getName();
+            //If read's mate is already in unpairedReads then we have a pair, which gets added to the ordered List
+            if (unpairedReads.containsKey(readName)) {
+                pairedReadsList.add(read);
+                pairedReadsList.add(unpairedReads.remove(readName));
+            } else {
+                unpairedReads.put(readName, read);
+            }
+        }
+        //Get the unpaired reads out of the hashmap
+        final List<GATKRead> unpairedReadsList = new ArrayList<>(unpairedReads.values());
+
+        //Minimize unpairedReads memory footprint (don't rely on readsPerPartitionGuess)
+        final List<GATKRead> pairedReadsListResized = new ArrayList<>(pairedReadsList.size());
```

Comment:
You could use the same `trimToSize` method on `ArrayList` we talked about in your previous PR.